pandas/core/window/rolling.py

```diff
@@ -93,6 +93,13 @@ def __init__(
         self.axis = obj._get_axis_number(axis) if axis is not None else None
         self.validate()
         self._numba_func_cache: Dict[Optional[str], Callable] = dict()
+        self.exclusions = kwargs.get("exclusions", set())
+
+    def _shallow_copy(self, obj: FrameOrSeries, **kwargs) -> ShallowMixin:
+        exclusions = self.exclusions
+        new_obj = super()._shallow_copy(obj, exclusions=exclusions, **kwargs)
+        new_obj.obj = new_obj._obj_with_exclusions
```

Comment:
are you sure this line is actually needed? this is very very odd to do

Reply:
copying the exclusions should be enough. you *maybe* be able to move this shallow copy to pandas/core/groupby/groupby.py which is better than here.

Reply:
The reason for doing this was that  `RollingGroupby._groupby._python_apply_general` operates on `RollingGroupby._groupby._selected_obj`.
And `RollingGroupby._groupby._selected_obj` returns `RollingGroupby._groupby.obj`, possibly sliced by `RollingGroupby._groupby._selection`

Reply:
> you maybe be able to move this shallow copy to pandas/core/groupby/groupby.py

As in, define `_GroupBy._shallow_copy`? I don't understand how that would work: when we call `Rolling.sum`, we go to `_Rolling_and_Expanding.sum` and then `WindowGroupByMixin._apply`. Inside there, we have
```
        def f(x, name=name, *args):
            x = self._shallow_copy(x)

            if isinstance(name, str):
                return getattr(x, name)(*args, **kwargs)

            return x.apply(name, *args, **kwargs)
```
Here, `self` is `WindowGroupByMixin`. In the case when we get here from `Rolling.sum`, then `self._shallow_copy` will take us directly to `ShallowMixin._shallow_copy` - even if we define `_GroupBy._shallow_copy`, we won't get there.

Reply:
@jreback thanks for your review. I've pushed a new attempt at a solution.

I patches `obj` _before_ we get to `_GroupBy.apply`, and only for `Rolling` so that previous behaviour in `DataFrame.groupby.apply`.